File: src/main/java/io/scriptor/conf/Node.java
package io.scriptor.conf;

import org.jetbrains.annotations.NotNull;

public interface Node<T> {

    default boolean has(final @NotNull String key) {
        return false;
    }

    default <N extends Node<?>> @NotNull N get(final @NotNull Class<N> type, final @NotNull String key) {
        throw new UnsupportedOperationException();
    }

    default <N extends Node<?>> @NotNull N get(final @NotNull Class<N> type, final int index) {
        throw new UnsupportedOperationException();
    }

    default @NotNull T value() {
        throw new UnsupportedOperationException();
    }
}

File: src/main/java/io/scriptor/conf/Node.kt
package io.scriptor.conf

import kotlin.reflect.KClass

interface Node<T> {
    operator fun contains(key: String): Boolean {
        return false
    }

    operator fun <N : Node<*>> get(type: KClass<N>, key: String): N {
        throw UnsupportedOperationException()
    }

    operator fun <N : Node<*>> get(type: KClass<N>, index: Int): N {
        throw UnsupportedOperationException()
    }

    fun value(): T {
        throw UnsupportedOperationException()
    }
}
